src/Microsoft.AspNetCore.SignalR.Protocols.MsgPack/Internal/Protocol/MessagePackHubProtocol.cs

```diff
@@ -65,39 +66,42 @@ public bool TryParseMessages(ReadOnlyMemory<byte> input, IInvocationBinder binde
 
         private static HubMessage ParseMessage(byte[] input, int startOffset, IInvocationBinder binder)
         {
-            using (var unpacker = Unpacker.Create(input, startOffset))
+            _ = MessagePackBinary.ReadArrayHeader(input, startOffset, out var readSize);
+            startOffset += readSize;
+            //using (var unpacker = Unpacker.Create(input, startOffset))
             {
-                _ = ReadArrayLength(unpacker, "elementCount");
+                //_ = ReadArrayLength(unpacker, "elementCount");
 
-                var messageType = ReadInt32(unpacker, "messageType");
+                var messageType = ReadInt32(input, ref startOffset, "messageType");
+                //var messageType = ReadInt32(unpacker, "messageType");
 
                 switch (messageType)
                 {
                     case HubProtocolConstants.InvocationMessageType:
-                        return CreateInvocationMessage(unpacker, binder);
+                        return CreateInvocationMessage(input, ref startOffset, binder);
                     case HubProtocolConstants.StreamInvocationMessageType:
-                        return CreateStreamInvocationMessage(unpacker, binder);
+                        return CreateStreamInvocationMessage(input, ref startOffset, binder);
                     case HubProtocolConstants.StreamItemMessageType:
-                        return CreateStreamItemMessage(unpacker, binder);
+                        return CreateStreamItemMessage(input, ref startOffset, binder);
                     case HubProtocolConstants.CompletionMessageType:
-                        return CreateCompletionMessage(unpacker, binder);
+                        return CreateCompletionMessage(input, ref startOffset, binder);
                     case HubProtocolConstants.CancelInvocationMessageType:
-                        return CreateCancelInvocationMessage(unpacker);
+                        return CreateCancelInvocationMessage(input, ref startOffset);
                     case HubProtocolConstants.PingMessageType:
                         return PingMessage.Instance;
                     case HubProtocolConstants.CloseMessageType:
-                        return CreateCloseMessage(unpacker);
+                        return CreateCloseMessage(input, ref startOffset);
```

Comment:
Does `startOffset` return how far into the byte array we read with msgpack? If we know that then I think we can optimizations to `BinaryMessageParser.TryParseMessage`.  We should no longer need to search for the next VarInt because we know exactly where it will start: immediately after the current message ends.

Could leave this for a separate follow up issue.

Reply:
@davidfowl

Reply:
In theory, we don't need the length prefix at all. MsgPack is fully self-describing (like JSON). The main reason we have the length prefix and the record separator is to allow us to determine if we have enough data to process the whole message without having to dive into the parser (only to figure out we don't have enough data to complete the object).

With the length prefix/record separator, we can use simple logic (reading a var int or scanning for a single byte) to quickly determine if we have a full message.